tty/usbacm/usbacm.c

```diff
@@ -117,91 +125,64 @@ static int usbacm_init(usbacm_dev_t *dev)
 }
 
 
-static void usbacm_rxthr(void *arg)
+static int usbacm_deinit(usbacm_dev_t *dev)
 {
-	usbacm_dev_t *dev = (usbacm_dev_t *)arg;
-	char buf[512];
-	int i, ret;
-
-	mutexLock(dev->fifoLock);
-	dev->rxRunning = 1;
-	dev->rxtid = gettid();
-	mutexUnlock(dev->fifoLock);
+	usb_setup_packet_t setup;
 
-	for (;;) {
-		if ((ret = usb_transferBulk(dev->pipeBulkIN, buf, sizeof(buf), usb_dir_in)) < 0)
-			break;
+	setup.bmRequestType = REQUEST_DIR_HOST2DEV | REQUEST_TYPE_CLASS | REQUEST_RECIPIENT_INTERFACE;
+	setup.bRequest = USB_CDC_REQ_SET_CONTROL_LINE_STATE;
+	setup.wIndex = dev->instance.interface;
+	setup.wValue = 0;
+	setup.wLength = 0;
 
-		mutexLock(dev->fifoLock);
-		for (i = 0; i < ret; i++) {
-			while (fifo_is_full(dev->fifo))
-				condWait(dev->fifoFull, dev->fifoLock, 0);
-			fifo_push(dev->fifo, buf[i]);
-		}
-		mutexUnlock(dev->fifoLock);
+	if (usb_transferControl(dev->pipeCtrl, &setup, NULL, 0, usb_dir_out) < 0) {
+		fprintf(stderr, "usbacm: Control transfer failed\n");
+		return -EIO;
 	}
 
-	mutexLock(dev->fifoLock);
-	dev->rxRunning = 0;
-	mutexUnlock(dev->fifoLock);
-
-	endthread();
+	return EOK;
 }
 
 
-static int _usbacm_readNonblock(usbacm_dev_t *dev, char *data, size_t len)
+static void usbacm_fifoPush(usbacm_dev_t *dev, char *data, size_t size)
 {
-	int full, i;
+	unsigned i;
+	int empty;
 
-	if (!dev->rxRunning)
-		return -EIO;
+	mutexLock(dev->fifoLock);
+	empty = fifo_is_empty(dev->fifo);
 
-	full = fifo_is_full(dev->fifo);
-	for (i = 0; i < len && !fifo_is_empty(dev->fifo); i++)
-		data[i] = fifo_pop_back(dev->fifo);
-	if (full)
-		condSignal(dev->fifoFull);
+	for (i = 0; i < size; i++)
+		fifo_push(dev->fifo, data[i]);
 
-	return i;
+	if (empty)
+		condSignal(dev->fifoEmpty);
+	mutexUnlock(dev->fifoLock);
```

Comment:
It's good idea to do condSignal outside of lock - condSignal does context switch